tools/flight_recorder/fr_trace.py

```diff
@@ -0,0 +1,831 @@
+                        id=len(nccl_calls),
+                        collective_id=None,
+                        group_id=pg_name,
+                        global_rank=r,
+                    )
+                )
+
+        if mismatch[pg_name] > MISMATCH_TAIL:
+            print(f"Too many mismatches for process_group {pg_name}:{desc}, aborting")
+            sys.exit(-1)
+
+    return tracebacks, collectives, nccl_calls
+
+
+def check_no_missing_dump_files(
+    entries: Dict[str, Any], memberships: List[Membership]
+) -> None:
+    all_ranks = set()
+    for membership in memberships:
+        all_ranks.add(str(membership.global_rank))
+    dumps_ranks = set(entries.keys())
+    assert (
+        dumps_ranks == all_ranks
+    ), f"Missing dump files from ranks {all_ranks - dumps_ranks}"
+
+
+def check_version(versions: Dict[str, Any]) -> None:
+    for rank, version in versions.items():  # noqa: PERF102
+        major, minor = map(int, version.split("."))
+        # assert major == 2, f"Rank {rank} unsupported version {version}"
+        # assert minor >= 0, f"Rank {rank} unsupported version {version}"
+
+
+def check_trace_from_beginning(entries: Dict[str, Any]) -> bool:
+    for rank in entries:
+        first_record_id = entries[rank][0]["record_id"]
+        # TODO add more sequence information such that analysis can proceed even without complete buffer
+
+        # assert first_record_id == 0, f"Rank {rank} trace does not start at time 0 (first record is {first_record_id}."
+        if first_record_id != 0:
+            print(
+                f"Rank {rank} trace does not start at time 0 (first record is {first_record_id}."
+            )
+            return False
+    return True
+
+
+def build_db(details: Dict[str, Dict[str, Any]], args: argparse.Namespace) -> Database:
+    # temporary state used for building database
+    entries = {}
+    pg_config = {}
+    version = {}
+    for dump in details.values():
+        rank = dump["rank"]
+        entries[rank] = dump["entries"]
+        version[rank] = dump["version"]
+        pg_config[rank] = dump["pg_config"]
+
+    check_version(version)
+    check_trace_from_beginning(entries)
+
+    # flattened database
+    groups, _groups, memberships, _memberships = build_groups_memberships(pg_config)
+    print("built groups, memberships")
+
+    check_no_missing_dump_files(entries, memberships)
+
+    if args.just_print_entries:
+        just_print_entries(entries, _groups, _memberships)
+        sys.exit(0)
+
+    tracebacks, collectives, nccl_calls = build_collectives(
+        entries, _groups, _memberships
+    )
+    print("built collectives, nccl_calls")
+    # print("Groups\n", tabulate(groups, headers=Group._fields))
+    # print("Memberships\n", tabulate(memberships, headers=Membership._fields))
+    # print("Collectives\n", tabulate(collectives, headers=Collective._fields))
+    # print("NCCLCalls\n", tabulate(nccl_calls, headers=NCCLCall._fields))
```

Comment:
Do we want to keep this?

Reply:
This is useful. I'll add a `--debug` switch and put all these behind the `debug` global switch.